packages/livekit-rtc/src/audio_stream.ts

```diff
@@ -70,15 +59,36 @@ export class AudioStream extends (EventEmitter as new () => TypedEmitter<AudioSt
     switch (streamEvent.case) {
       case 'frameReceived':
         const frame = AudioFrame.fromOwnedInfo(streamEvent.value.frame);
-        this.emit(AudioStreamEvent.FrameReceived, { frame });
+        if (this.queueResolve) {
+          this.queueResolve({ done: false, value: frame });
+        } else {
+          this.eventQueue.push(frame);
+        }
         break;
       case 'eos':
         FfiClient.instance.off(FfiClientEvent.FfiEvent, this.onEvent);
         break;
     }
   };
 
+  async next(): Promise<IteratorResult<AudioFrame>> {
+    if (this.eventQueue.length > 0) {
+      const value = this.eventQueue.shift();
+      if (value) {
+        return { done: false, value };
+      } else {
+        return { done: true, value: undefined };
+      }
+    }
+    return new Promise((resolve) => (this.queueResolve = resolve));
```

Comment:
hm. I think `this.queueResolve` might get overridden before being called?
so the async iterator could get stuck

Reply:
the `Mutex` helper that Théo has mentioned already might make sense here

Reply:
could it? i don't think that's the case, it only gets overridden if you call `next()` twice without awaiting the first one, which you shouldn't do anyway.

Reply:
> which you shouldn't do anyway.

agreed that it _shouldn't_ be done, but seems like a footgun regardless?

Reply:
would it make sense to return a copy of the same promise if `next` is called twice?

Reply:
I'm not sure, I think best case scenario is a promise chain, which is exactly what the `Mutex` class does.